File: javascript/src/credits/types/analytics.ts
import type { Decimal } from "decimal.js";
import type { ListUsageChargesOptions, UsageChargePage } from "./ledger.js";

// ── Usage analytics ─────────────────────────────────────────────────
/** Aggregated spend for a single user in a time window. */
export interface SpendByUserRow {
  userId: string;
  totalSpend: Decimal;
  entryCount: number;
}

/** Aggregated spend for a single model in a time window. */
export interface SpendByModelRow {
  model: string;
  totalSpend: Decimal;
  entryCount: number;
}

/** Top-spending user in a time window. */
export interface TopUserRow {
  userId: string;
  totalSpend: Decimal;
}

/** Daily spend aggregation in a time window. */
export interface DailySpendRow {
  date: string;
  totalSpend: Decimal;
  entryCount: number;
}

/** Aggregate statistics across all users in a time window. */
export interface AggregateStats {
  totalCreditsConsumed: Decimal;
  activeUsers: number;
  avgDailySpend: Decimal;
  topModel: string | null;
  topUser: string | null;
}

/**
 * Read-only usage analytics backend.
 *
 * PostgreSQL implements this contract by default. High-volume deployments can
 * provide a ClickHouse implementation without moving balances or compact
 * accounting receipts out of PostgreSQL.
 */
export interface UsageAnalyticsStore {
  spendByUser(start: Date, end: Date): Promise<SpendByUserRow[]>;
  spendByModel(start: Date, end: Date): Promise<SpendByModelRow[]>;
  topUsers(limit: number, start: Date, end: Date): Promise<TopUserRow[]>;
  dailySpend(start: Date, end: Date): Promise<DailySpendRow[]>;
  aggregateStats(start: Date, end: Date): Promise<AggregateStats>;
}

/** Read-only usage history backend selected with the analytics backend. */
export interface UsageChargeStore {
  listUsageCharges(userId: string, options?: ListUsageChargesOptions): Promise<UsageChargePage>;
}

File: python/src/bursar/credits/types/analytics.py
"""Analytics types — mirrors JS SDK's ``credits/types/analytics.ts``."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Protocol, runtime_checkable

from pydantic import BaseModel

from bursar.credits.types.ledger import UsageChargeCursor, UsageChargePage


class SpendByUserRow(BaseModel):
    user_id: str
    total_spend: Decimal
    entry_count: int


class SpendByModelRow(BaseModel):
    model: str
    total_spend: Decimal
    entry_count: int


class TopUserRow(BaseModel):
    user_id: str
    total_spend: Decimal


class DailySpendRow(BaseModel):
    date: str
    total_spend: Decimal
    entry_count: int


class AggregateStats(BaseModel):
    total_credits_consumed: Decimal
    active_users: int
    avg_daily_spend: Decimal
    top_model: str
    top_user: str


@runtime_checkable
class UsageAnalyticsStore(Protocol):
    """Read-only usage analytics backend.

    PostgreSQL implements this protocol by default. High-volume deployments
    can provide ClickHouse without moving balances or compact accounting
    receipts out of PostgreSQL.
    """

    def spend_by_user(self, start: datetime, end: datetime) -> list[SpendByUserRow]: ...

    def spend_by_model(self, start: datetime, end: datetime) -> list[SpendByModelRow]: ...

    def top_users(self, limit: int, start: datetime, end: datetime) -> list[TopUserRow]: ...

    def daily_spend(self, start: datetime, end: datetime) -> list[DailySpendRow]: ...

    def aggregate_stats(self, start: datetime, end: datetime) -> AggregateStats: ...


@runtime_checkable
class UsageChargeStore(Protocol):
    """Read-only usage history backend selected with the analytics backend."""

    def list_usage_charges(
        self,
        user_id: str,
        from_date: datetime | None = None,
        to_date: datetime | None = None,
        limit: int = 50,
        cursor: UsageChargeCursor | None = None,
        include_record_only: bool = True,
    ) -> UsageChargePage: ...
